File: Backend/project/static/js/class/disbursement.js
export class Desembolso {
    constructor(
        credit_id = 0,
        monto_total_desembolso = 0,
        honorarios = 0,
        poliza_seguro = 0,
        monto_credito = 0,
        saldo_anterior = 0,
        forma_desembolso = 'APLICACIÓN GASTOS'
    ) {
        this._credit_id = credit_id;
        this._monto_total_desembolso = monto_total_desembolso;
        this._honorarios = honorarios;
        this._poliza_seguro = poliza_seguro;
        this._monto_credito = monto_credito;
        this._saldo_anterior = saldo_anterior;
        this._forma_desembolso = forma_desembolso;
    }

    // Métodos get
    get credit_id() {
        return this._credit_id;
    }

    get monto_total_desembolso() {
        return this._monto_total_desembolso;
    }

    get honorarios() {
        return this._honorarios;
    }

    get poliza_seguro() {
        return this._poliza_seguro;
    }

    get monto_credito() {
        return this._monto_credito;
    }

    get saldo_anterior() {
        return this._saldo_anterior;
    }

    get forma_desembolso() {
        return this._forma_desembolso;
    }

    // Métodos set
    set credit_id(value) {
        this._credit_id = value;
    }

    set monto_total_desembolso(value) {
        this._monto_total_desembolso = value;
    }

    set honorarios(value) {
        this._honorarios = value;
    }

    set poliza_seguro(value) {
        this._poliza_seguro = value;
    }

    set monto_credito(value) {
        this._monto_credito = value;
    }

    set saldo_anterior(value) {
        this._saldo_anterior = value;
    }

    set forma_desembolso(value) {
        this._forma_desembolso = value;
    }

    // Método para calcular el total a depositar
    get total_a_depositar() {
        return this._monto_credito - (this._honorarios + this._poliza_seguro + this._saldo_anterior);
    }

    // Método para convertir a JSON
    toJson() {
        return {
            forma_desembolso: this._forma_desembolso,
            monto_credito: this._monto_credito,
            saldo_anterior: this._saldo_anterior,
            honorarios: this._honorarios,
            poliza_seguro: this._poliza_seguro,
            monto_total_desembolso: this.total_a_depositar, // Sin paréntesis
            credit_id: this._credit_id
        };
    }
}

File: Backend/project/apps/financings/models/disbursement.py
from django.db import models

# MODELO
from .credit import Credit

# FORMATO
from apps.financings.formato import formatear_numero

# DESEMBOLSO
class Disbursement(models.Model):
    formaDesembolso = [
        ('APLICACIÓN GASTOS', 'APLICACIÓN GASTOS'),
        ('APLICACIÓN DE AMPLIACIÓN DE CRÉDITO VIGENTE', 'APLICACIÓN DE AMPLIACIÓN DE CRÉDITO VIGENTE'),
        ('CANCELACIÓN DE CRÉDITO VIGENTE', 'CANCELACIÓN DE CRÉDITO VIGENTE'),
        ('DESEMBOLSAR', 'DESEMBOLSAR')
    ]
    credit_id = models.ForeignKey(Credit, on_delete=models.CASCADE, verbose_name='Credito')
    forma_desembolso = models.CharField("Forma de Desembolso", choices=formaDesembolso, max_length=75, blank=False, null=False)
    monto_credito = models.DecimalField("Monto Credito", decimal_places=2, max_digits=15, default=0)
    monto_credito_agregar = models.DecimalField("Monto Credito Agregar", decimal_places=2, max_digits=15, default=0)
    monto_credito_cancelar = models.DecimalField("Monto Credito Cancelar", decimal_places=2, max_digits=15, default=0) 
    saldo_anterior = models.DecimalField("Saldo Anterior", decimal_places=2, max_digits=15, default=0)
    honorarios = models.DecimalField("Honorarios", decimal_places=2, max_digits=15, default=0)
    poliza_seguro = models.DecimalField("Poliza de Seguro", decimal_places=2, max_digits=15, default=0)
    monto_desembolsado = models.DecimalField("Monto Desembolsado", decimal_places=2, max_digits=15, default=0)
    total_gastos  = models.DecimalField("Total de gastos", decimal_places=2, max_digits=15, default=0)
    monto_total_desembolso = models.DecimalField("Monto Total a Desembolsar", decimal_places=2, max_digits=15, default=0)
    total_t = models.DecimalField("Total de totales", decimal_places=2, max_digits=15, default=0)
    description = models.TextField("Descripcion", blank=True, null=True)
    creation_date = models.DateTimeField("Fecha de Creación", auto_now_add=True)
    #Nuevos atributos
    #credito_cancelado = models.ForeignKey(Credit, on_delete=models.CASCADE, verbose_name='Credito', blank=True, null=True)

    
    def f_monto_credito(self):
        return formatear_numero(self.monto_credito)

    def f_monto_credito_agregar(self):
        return formatear_numero(self.monto_credito_agregar)
    
    def f_monto_credito_cancelar(self):
        return formatear_numero(self.monto_credito_cancelar)
    
    def f_saldo_anterior(self):
        return formatear_numero(self.saldo_anterior)
    
    def f_honorarios(self):
        return formatear_numero(self.honorarios)
    
    def f_poliza_seguro(self):
        return formatear_numero(self.poliza_seguro)
    
    def f_monto_desembolsado(self):
        return formatear_numero(self.monto_desembolsado)
    
    def f_total_gastos(self):
        return formatear_numero(self.total_gastos)
    
    def f_monto_total_desembolso(self):
        return formatear_numero(self.monto_total_desembolso)

    def __str__(self):
        return f'{self.forma_desembolso}'

    class Meta:
        verbose_name = "Desembolso"
        verbose_name_plural = "Desembolsos"